spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubHealthIndicator.java

```diff
@@ -16,55 +16,171 @@
+ * @author Patrik Hörlin
  *
  * @since 1.2.2
  */
 public class PubSubHealthIndicator extends AbstractHealthIndicator {
 
+	/**
+	 * Template used when performing health check calls.
+	 */
 	private final PubSubTemplate pubSubTemplate;
 
-	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate) {
+	/**
+	 * Indicates whether a user subscription has been configured.
+	 */
+	private final boolean specifiedSubscription;
+
+	/**
+	 * Subscription used when health checking.
+	 */
+	private final String subscription;
+
+	/**
+	 * Timeout when performing health check.
+	 */
+	private final long timeoutMillis;
+
+	/**
+	 * If messages pull should be acknowledged or not.
+	 */
+	private final boolean acknowledgeMessages;
+
+	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate, String healthCheckSubscription, long timeoutMillis, boolean acknowledgeMessages) {
 		super("Failed to connect to Pub/Sub APIs. Check your credentials and verify you have proper access to the service.");
-		Assert.notNull(pubSubTemplate, "PubSubTemplate can't be null");
+		Assert.notNull(pubSubTemplate, "pubSubTemplate can't be null");
 		this.pubSubTemplate = pubSubTemplate;
+		this.specifiedSubscription = StringUtils.hasText(healthCheckSubscription);
+		if (this.specifiedSubscription) {
+			this.subscription = healthCheckSubscription;
+		}
+		else {
+			this.subscription = UUID.randomUUID().toString();
+		}
+		this.timeoutMillis = timeoutMillis;
+		this.acknowledgeMessages = acknowledgeMessages;
+	}
+
+	void validateHealthCheck() {
+		doHealthCheck(
+				() -> { },
+				this::validationFailed,
+				this::validationFailed);
 	}
 
 	@Override
 	protected void doHealthCheck(Health.Builder builder) {
+		doHealthCheck(
+				builder::up,
+				builder::down,
+				e -> builder.withException(e).unknown());
+	}
+
+	private void doHealthCheck(Runnable up, Consumer<Throwable> down, Consumer<Throwable> unknown) {
 		try {
-			this.pubSubTemplate.pull("subscription-" + UUID.randomUUID().toString(), 1, true);
+			pullMessage();
+			up.run();
 		}
-		catch (ApiException aex) {
-			Code errorCode = aex.getStatusCode().getCode();
-			if (errorCode == StatusCode.Code.NOT_FOUND || errorCode == Code.PERMISSION_DENIED) {
-				builder.up();
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			unknown.accept(e);
+		}
+		catch (ExecutionException e) {
+			if (isHealthyException(e)) {
+				up.run();
 			}
 			else {
-				builder.withException(aex).down();
+				down.accept(e);
 			}
 		}
+		catch (TimeoutException e) {
+			unknown.accept(e);
+		}
 		catch (Exception e) {
-			builder.withException(e).down();
+			down.accept(e);
+		}
+	}
+
+	private void pullMessage() throws InterruptedException, ExecutionException, TimeoutException {
+		ListenableFuture<List<AcknowledgeablePubsubMessage>> future = pubSubTemplate.pullAsync(this.subscription, 1, true);
+		List<AcknowledgeablePubsubMessage> messages = future.get(timeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
I think using the `future.get(..)` is blocking right? Should you register a callback here?

Reply:
We decided to keep it synchronous.

Reply:
I see. Is there a benefit to using pullAsync vs. the Sync pull then?